OrbitQt/servicedeploymanager.cpp

```diff
@@ -166,70 +167,78 @@ outcome::result<void> ServiceDeployManager::CopyFileToRemote(
   return outcome::success();
 }
 
-outcome::result<void> ServiceDeployManager::StopSftpChannel(
-    OrbitSshQt::SftpChannel* channel) {
-  auto quit_handler =
-      ConnectQuitHandler(channel, &OrbitSshQt::SftpChannel::stopped);
+void ServiceDeployManager::StopSftpChannel() {
+  auto quit_handler = ConnectQuitHandler(&sftp_channel_.value(),
+                                         &OrbitSshQt::SftpChannel::stopped);
 
-  auto error_handler =
-      ConnectErrorHandler(channel, &OrbitSshQt::SftpChannel::errorOccurred);
+  auto error_handler = ConnectErrorHandler(
+      &sftp_channel_.value(), &OrbitSshQt::SftpChannel::errorOccurred);
 
-  channel->Stop();
+  sftp_channel_->Stop();
 
-  OUTCOME_TRY(loop_.exec());
-  return outcome::success();
+  (void)loop_.exec();
 }
 
 outcome::result<void> ServiceDeployManager::CopyOrbitServicePackage() {
   emit statusMessage("Copying OrbitService package to the remote instance...");
 
-  OrbitSshQt::SftpChannel channel{&session_.value()};
-
-  OUTCOME_TRY(StartSftpChannel(&channel));
-  LOG("SFTP Channel open.");
-
   auto& config =
       std::get<SignedDebianPackageDeployment>(deployment_configuration_);
 
   using FileMode = OrbitSshQt::SftpCopyToRemoteOperation::FileMode;
 
   OUTCOME_TRY(
-      MapError(CopyFileToRemote(&channel, config.path_to_package.string(),
+      MapError(CopyFileToRemote(config.path_to_package.string(),
                                 kDebDestinationPath, FileMode::kUserWritable),
                Error::kCouldNotUploadPackage));
 
   OUTCOME_TRY(
-      MapError(CopyFileToRemote(&channel, config.path_to_signature.string(),
+      MapError(CopyFileToRemote(config.path_to_signature.string(),
                                 kSigDestinationPath, FileMode::kUserWritable),
                Error::kCouldNotUploadSignature));
 
-  OUTCOME_TRY(StopSftpChannel(&channel));
-
   emit statusMessage(
       "Finished copying the OrbitService package to the remote instance.");
   return outcome::success();
 }
 
+ErrorMessageOr<void> ServiceDeployManager::CopyFileToLocal(
+    std::string_view source, std::string_view destination) {
+  LOG("Copying remote \"%s\" to local \"%s\"", source, destination);
+  OrbitSshQt::SftpCopyToLocalOperation operation{&session_.value(),
+                                                 &sftp_channel_.value()};
+
+  auto quit_handler = ConnectQuitHandler(
+      &operation, &OrbitSshQt::SftpCopyToLocalOperation::stopped);
+
+  auto error_handler = ConnectErrorHandler(
+      &operation, &OrbitSshQt::SftpCopyToLocalOperation::errorOccurred);
+
+  operation.CopyFileToLocal(source, destination);
+
+  auto result = loop_.exec();
```

Comment:
Here is some kind of problem.

When symbols of two (or) more modules that are not in the cache are requested at the same time, then this error is printed:`QEventLoop::exec: instance 0x7ffcbf53ae60 has already called exec()` and the loading of any but one of the files fails. 
This situation can occur when the user selects several modules and right clicks load symbols and if a preset is loaded.

Reply:
Oh well, I guess it forces us back into callbacks world. @beckerhe is there another way this use-case can be handled? I do not think we can create another instance of an event loop since we have a lot of ssh related signals set up in this one.

Reply:
In this case don't use the member "loop_". When you create a local `QEventLoop`-instance in that function it should work. But you can't use "ConnectQuitHandler" and "ConnectErrorHandler" though since these connect to `loop_`. You probably should overload those with a version that takes a QEventLoop-pointer.

Reply:
ok, since this change is actually doing a lot I am going to fixup and rebase everything. Please bare with me and take another look after I force-push next revision.

Reply:
done.